scipy/stats/tests/test_rank.py

```diff
@@ -73,63 +72,115 @@ def test_overflow(self):
         assert_equal(out, 1.0 - k * (ntie**3 - ntie) / float(n**3 - n))
 
 
+_rankdata_cases = (
+    # values, method, expected
+    ([], 'average', []),
+    ([], 'min', []),
+    ([], 'max', []),
+    ([], 'dense', []),
+    ([], 'ordinal', []),
+    #
+    ([100], 'average', [1.0]),
+    ([100], 'min', [1.0]),
+    ([100], 'max', [1.0]),
+    ([100], 'dense', [1.0]),
+    ([100], 'ordinal', [1.0]),
+    #
+    ([100, 100, 100], 'average', [2.0, 2.0, 2.0]),
+    ([100, 100, 100], 'min', [1.0, 1.0, 1.0]),
+    ([100, 100, 100], 'max', [3.0, 3.0, 3.0]),
+    ([100, 100, 100], 'dense', [1.0, 1.0, 1.0]),
+    ([100, 100, 100], 'ordinal', [1.0, 2.0, 3.0]),
+    #
+    ([100, 300, 200], 'average', [1.0, 3.0, 2.0]),
+    ([100, 300, 200], 'min', [1.0, 3.0, 2.0]),
+    ([100, 300, 200], 'max', [1.0, 3.0, 2.0]),
+    ([100, 300, 200], 'dense', [1.0, 3.0, 2.0]),
+    ([100, 300, 200], 'ordinal', [1.0, 3.0, 2.0]),
+    #
+    ([100, 200, 300, 200], 'average', [1.0, 2.5, 4.0, 2.5]),
+    ([100, 200, 300, 200], 'min', [1.0, 2.0, 4.0, 2.0]),
+    ([100, 200, 300, 200], 'max', [1.0, 3.0, 4.0, 3.0]),
+    ([100, 200, 300, 200], 'dense', [1.0, 2.0, 3.0, 2.0]),
+    ([100, 200, 300, 200], 'ordinal', [1.0, 2.0, 4.0, 3.0]),
+    #
+    ([100, 200, 300, 200, 100], 'average', [1.5, 3.5, 5.0, 3.5, 1.5]),
+    ([100, 200, 300, 200, 100], 'min', [1.0, 3.0, 5.0, 3.0, 1.0]),
+    ([100, 200, 300, 200, 100], 'max', [2.0, 4.0, 5.0, 4.0, 2.0]),
+    ([100, 200, 300, 200, 100], 'dense', [1.0, 2.0, 3.0, 2.0, 1.0]),
+    ([100, 200, 300, 200, 100], 'ordinal', [1.0, 3.0, 5.0, 4.0, 2.0]),
+    #
+    ([10] * 30, 'ordinal', np.arange(1.0, 31.0)),
+)
+
+
+@pytest.mark.skip_xp_backends('torch', reason='no repeat')
 class TestRankData:
 
-    def test_empty(self):
-        """stats.rankdata([]) should return an empty array."""
-        a = np.array([], dtype=int)
+    def desired_dtype(self, method='average', has_nans=False, *, xp):
+        if has_nans:
+            return xp.asarray(1.).dtype
+        return xp.asarray(1.).dtype if method=='average' else xp.asarray(1).dtype
+
+    def test_empty(self, xp):
+        """stats.rankdata of empty array should return an empty array."""
+        a = xp.asarray([], dtype=xp.int64)
         r = rankdata(a)
-        assert_array_equal(r, np.array([], dtype=np.float64))
-        r = rankdata([])
-        assert_array_equal(r, np.array([], dtype=np.float64))
+        xp_assert_equal(r, xp.asarray([], dtype=self.desired_dtype(xp=xp)))
+        if is_numpy(xp):
+            r = rankdata([])
+            assert_array_equal(r, np.array([], self.desired_dtype(xp=xp)))
```

Comment:
If I had translated this today, I probably wouldn't have special-cased NumPy. Maybe leave this out of each test, and do just one test separately.